packages/core/primitives/event-dispatch/src/event_dispatcher.ts

```diff
@@ -0,0 +1,153 @@
+    [PROPAGATION_STOPPED_SYMBOL]?: boolean;
+  }
+}
+
+/**
+ * A dispatcher that uses browser-based `Event` semantics, for example bubbling, `stopPropagation`,
+ * `currentTarget`, etc.
+ */
+export class EventDispatcher {
+  private readonly actionResolver: ActionResolver;
+
+  private readonly dispatcher: Dispatcher;
+
+  constructor(private readonly dispatchDelegate: (event: Event, actionName: string) => void) {
+    this.actionResolver = new ActionResolver();
+    this.dispatcher = new Dispatcher(
+      (eventInfoWrapper: EventInfoWrapper) => {
+        this.dispatchToDelegate(eventInfoWrapper);
+      },
+      {
+        actionResolver: this.actionResolver,
+        eventReplayer: (eventInfoWrappers: EventInfoWrapper[]) => {
+          for (const eventInfoWrapper of eventInfoWrappers) {
+            prepareEventForReplay(eventInfoWrapper);
+            this.dispatchToDelegate(eventInfoWrapper);
+          }
+        },
+      },
+    );
+  }
+
+  /**
+   * The entrypoint for the `EventContract` dispatch.
+   */
+  dispatch(eventInfo: EventInfo): void {
+    this.dispatcher.dispatch(eventInfo);
+  }
+
+  /** Internal method that does basic disaptching. */
+  private dispatchToDelegate(eventInfoWrapper: EventInfoWrapper) {
+    prepareEventForBubbling(eventInfoWrapper);
+    while (eventInfoWrapper.getAction()) {
+      prepareEventForDispatch(eventInfoWrapper);
+      this.dispatchDelegate(eventInfoWrapper.getEvent(), eventInfoWrapper.getAction()!.name);
+      if (propagationStopped(eventInfoWrapper)) {
+        return;
+      }
+      this.actionResolver.resolveParentAction(eventInfoWrapper.eventInfo);
+    }
+  }
+}
+
+function prepareEventForBubbling(eventInfoWrapper: EventInfoWrapper) {
+  const event = eventInfoWrapper.getEvent();
+  const stopPropagation = () => {
+    event[PROPAGATION_STOPPED_SYMBOL] = true;
+  };
+  Object.defineProperty(event, 'stopPropagation', {
+    value: stopPropagation,
+  });
+  Object.defineProperty(event, 'stopImmediatePropagation', {
+    value: stopPropagation,
+  });
+}
+
+function propagationStopped(eventInfoWrapper: EventInfoWrapper) {
+  const event = eventInfoWrapper.getEvent();
+  return !!event[PROPAGATION_STOPPED_SYMBOL];
+}
+
+function prepareEventForReplay(eventInfoWrapper: EventInfoWrapper) {
+  const event = eventInfoWrapper.getEvent();
+  const target = eventInfoWrapper.getTargetElement();
+  Object.defineProperty(event, 'target', {
+    value: target,
+  });
+  Object.defineProperty(event, 'eventPhase', {
+    value: EventPhase.REPLAY,
+  });
+  if (ngDevMode) {
+    Object.defineProperty(event, 'preventDefault', {
+      value: () => {
+        throw new Error(
+          '`preventDefault` called during event replay. Because event replay occurs after browser ' +
+            'dispatch, `preventDefault` would have no effect. You can check whether an event is ' +
+            'being replayed by accessing the event phase: `event.eventPhase === EventPhase.REPLAY`.',
+        );
+      },
```

Comment:
As we've discussed offline, we can patch those fields in both dev and prod modes, but tree-shake-away a large part of the error message, for example:
```suggestion
      value: () => {
        const details = ngDevMode ? 'Because event replay occurs after browser ' +
              'dispatch, `preventDefault` would have no effect. You can check whether an event is ' +
              'being replayed by accessing the event phase: `event.eventPhase === EventPhase.REPLAY`.'
            : '';
        throw new Error('`preventDefault` called during event replay. ${details}');
      },
```

Reply:
Done. I moved the messages and details to global constants, though.